---
output:
  pdf_document: default
  html_document: default
---
---Charles Walker Cano
title: "Transcriptomalysis "
author: "T. Van Laar, PhD. Charles Walker Cano"
date: "`r Sys.Date(5-7-24)`"
output: pdf_document
---

# Load required packages (you might have to figure out how to install some of these first...)
```{r message=FALSE}
library(ballgown)
library(RColorBrewer)
library(genefilter)
library(dplyr)
library(devtools)
```

# What is this code doing?
  This code is creating a specimen id column and a column of values that you specify. 
```{r}
pheno_data<-data.frame(ids = c("plank01", "plank02", "biofilm01", "biofilm02"),
                       stage = c("planktonic", "planktonic", "biofilm", "biofilm"))
```

# create Ballgown object and check transcript number
```{r message=FALSE}
samples.c <- paste('ballgown', pheno_data$ids, sep = '/')
bg <- ballgown(samples = samples.c, meas='all', pData = pheno_data)
bg
```

# <what is this code doing?>
Filters and displays data frames greater than 1.
```{r}
bg_filt = subset(bg,"rowVars(texpr(bg)) >1",genomesubset=TRUE)
bg_filt
```

# create a table of transcripts
```{r}
results_transcripts<- stattest(bg_filt, feature = "transcript", covariate = "stage",
getFC = TRUE, meas = "FPKM")
results_transcripts<-data.frame(geneNames=geneNames(bg_filt),
transcriptNames=transcriptNames(bg_filt), results_transcripts)
```

# choose a transcript to examine more closely (this is a demo, you need to choose another)
```{r}
results_transcripts[results_transcripts$transcriptNames == "gene-PA0004", ]
```
#Transcript that is examined at a closer look is gene-PA0004. 
The transcripts yields data of the gene name, feature, id# and fc value of 1.735612e+15.


## what information are you given about this transcript?
The transcripts yields data of the gene name, feature, id# and fc value of 1.735612e+15.

# <insert brief explanation of this code>
This code is filtering rows with a p-value that is less than 0.05. This means that with the product of the filtered rows they will all be significant. 
```{r}
sigdiff <- results_transcripts %>% filter(pval<0.05)
dim(sigdiff)
```

# organize the table <by what metrics is the table being organized?>
The matrix that is being used is matrix pva1 column in a increasing order within each identical group. 

```{r}
o = order(sigdiff[,"pval"], -abs(sigdiff[,"fc"]), decreasing=FALSE)
output = sigdiff[o,c("geneNames","transcriptNames", "id","fc","pval","qval")]
write.table(output, file="SigDiff.txt", sep="\t", row.names=FALSE, quote=FALSE)
head(output)
```

# load gene names
```{r}
bg_table = texpr(bg_filt, 'all')
bg_gene_names = unique(bg_table[, 9:10])
```

# pull out gene expression data and visualize
```{r}
gene_expression = as.data.frame(gexpr(bg_filt))
head(gene_expression)
```


# <what is this code doing? hint:compare the above output of head(gene_expression) to this output>

This code filters and assigns new columns to the gene expression as well as the demensions of the data frame.


```{r}
colnames(gene_expression) <- c("plank01", "plank02", "biofilm01", "biofilm02")
head(gene_expression)
dim(gene_expression)
```

# load the transcript to gene table and determine the number of transcripts and unique genes
```{r}
transcript_gene_table = indexes(bg)$t2g
head(transcript_gene_table)
length(row.names(transcript_gene_table))
length(unique(transcript_gene_table[,"g_id"]))
```

# plot the number of transcripts per gene
```{r}
counts=table(transcript_gene_table[,"g_id"])
c_one = length(which(counts == 1))
c_more_than_one = length(which(counts > 1))
c_max = max(counts)
hist(counts, breaks=50, col="red", xlab="Transcripts per gene",
main="Distribution of transcript count per gene")
legend_text = c(paste("Genes with one transcript =", c_one),
paste("Genes with more than one transcript =", c_more_than_one),
paste("Max transcripts for single gene = ", c_max))
legend("topright", legend_text, lty=NULL)
```

The graph above is plotting the number of transcripts per gene. The majority of the transcripts are taking place 1-2.



# create a plot of how similar the two replicates are for one another. We have two data sets...how can you modify this code in another chunk to create a plot of the other set?

The code can be modified by changing x = gene_expression[,"plank01"]
y = gene_expression[,"plank02"] to x = gene_expression[,"biofilm01"]
y = gene_expression[,"biofilm02"]. This will allow you to display two graphs to compare. 

```{r}
x = gene_expression[,"plank01"]
y = gene_expression[,"plank02"]
min_nonzero=1
plot(x=log2(x+min_nonzero), y=log2(y+min_nonzero), pch=16, col="red", cex=0.25,
xlab="FPKM (Planktonic, Replicate 1)", ylab="FPKM (Planktonic, Replicate 2)",
main="Comparison of expression values for a pair of replicates")
abline(a=0,b=1, col = "red")
rs=cor(x,y)^2
legend("topleft", paste("R squared = ", round(rs, digits=3), sep=""), lwd=1, col="gold")
```

```{r}
x = gene_expression[,"biofilm01"]
y = gene_expression[,"biofilm02"]
min_nonzero=1
plot(x=log2(x+min_nonzero), y=log2(y+min_nonzero), pch=16, col="red", cex=0.25,
xlab="FPKM (Biofilm, Replicate 1)", ylab="FPKM (Biofilm, Replicate 2)",
main="Comparison of expression values for a pair of replicates")
abline(a=0,b=1, col = "hotpink")
rs=cor(x,y)^2
legend("topleft", paste("R squared = ", round(rs, digits=3), sep=""), lwd=1, col="orange")
```



## What does it mean if the two data sets are similar?

When two datasets are similiar they exhibit the same patterns or characteristics that are comparable.

# create plot of differential gene expression between the conditions
```{r}
results_genes = stattest(bg_filt, feature="gene", covariate="stage", getFC=TRUE, meas="FPKM")
results_genes = merge(results_genes,bg_gene_names,by.x=c("id"),by.y=c("gene_id"))
sig=which(results_genes$pval<0.05)
results_genes[,"de"] = log2(results_genes[,"fc"])
hist(results_genes[sig,"de"], breaks=50, col="seagreen",
xlab="log2(Fold change) Planktonic vs Biofilm",
main="Distribution of differential expression values")
abline(v=-2, col="black", lwd=2, lty=2)
abline(v=2, col="black", lwd=2, lty=2)
legend("topleft", "Fold-change > 4", lwd=2, lty=2)
```

## interpret the above figure:
The graph is showing a histogram that represents distribution of the differential expression values between the biofilm and planktonic sample stages. The samples shown are expressed because they are calculated as being significant. This will aid researchers in looking at magnitude and direction of gene expression between the two stages. 



# Plot total gene expression highlighting differentially expressed genes
```{r}
gene_expression[,"plank"]=apply(gene_expression[,c(1:2)], 1, mean)
gene_expression[,"biofilm"]=apply(gene_expression[,c(3:4)], 1, mean)
x=log2(gene_expression[,"plank"]+min_nonzero)
y=log2(gene_expression[,"biofilm"]+min_nonzero)
plot(x=x, y=y, pch=16, cex=0.25, xlab="Planktonic FPKM (log2)", ylab="Biofilm FPKM (log2)",
main="Planktonic vs Biofilm FPKMs")
abline(a=0, b=1)
xsig=x[sig]
ysig=y[sig]
points(x=xsig, y=ysig, col="magenta", pch=16, cex=0.5)
legend("topleft", "Significant", col="magenta", pch=16)
```

# make a table of FPKM values
```{r}
fpkm = texpr(bg_filt,meas="FPKM")
```

# choose a gene to determine individual expression (pick a different number than I did)
```{r}
ballgown::transcriptNames(bg_filt)[24]
ballgown::geneNames(bg_filt)[24]
```
# transform to log2
```{r}
transformed_fpkm <- log2(fpkm[24, ] + 1)
```

# make sure values are properly coded as numbers
```{r}
numeric_stages <- as.numeric(factor(pheno_data$stage))

jittered_stages <- jitter(numeric_stages)
```

# plot expression of individual gene
```{r}
boxplot(transformed_fpkm ~ pheno_data$stage,
        main=paste(ballgown::geneNames(bg_filt)[24], ' : ', ballgown::transcriptNames(bg_filt)[24]),
        xlab="Stage",
        ylab="log2(FPKM+1)",
        col=c("red", "gold"),
        border="black")

points(transformed_fpkm ~ jittered_stages, 
       pch=21, col="blue", bg="lightblue", cex=1.2)
```

## interpret the above figure
The graph above is visualizing gene expression at different stages for biofilm and planktonic with an emphasis on the gene that I selected which was gene 24 (aroE/gene-PA0025).

